File: Stack-Queue-LinkedList/Deque/Deque.java
package StackQueueLinkedList;
import  java.util.*;
//deque using stacks

public class Deque {
 Stack<Integer> stack1;
 Stack<Integer> stack2;
 Stack<Integer> stack3;
 public Deque(){
     Stack<Integer> st1 = new Stack<>();
     Stack<Integer> st2 = new Stack<>();
     Stack<Integer> st3 = new Stack<>();
     stack1 = st1;
     stack2 = st2;
     stack3 = st3;
 }

 public void lPush(int x) {
     stack1.push(x);
 }

  public void rPush(int x) {
     stack2.push(x);
 }

  public int lPop() {
	 if(this.isEmpty()) {
		 throw new EmptyStackException();
	 }
     if(stack1.empty()) {
         int size = stack2.size()/2;
         while(size>0) {
             stack3.push(stack2.pop());
             size--;
         }
         while(!stack2.empty()) {
             stack1.push(stack2.pop());
         }
         while(!stack3.empty()) {
             stack2.push(stack3.pop());
         }
     }
     return stack1.pop();
 }

  public int rPop() {
	 if(this.isEmpty()) {
		 throw new EmptyStackException();
	 }
     if(stack2.empty()) {
         int size = stack1.size()/2;
         while(size>0) {
             stack3.push(stack1.pop());
             size--;
         }
         while(!stack1.empty()) {
             stack2.push(stack1.pop());
         }
         while(!stack3.empty()) {
             stack1.push(stack3.pop());
         }
     }
     return stack2.pop();
 }
  
  public int size() {
	  return stack1.size() + stack2.size();
  }
  
  public boolean isEmpty() {
	  return stack1.empty() && stack2.empty();
  }
  
  public static void main(String []args){
	  Deque d = new Deque();
	  d.rPush(4);
	  d.rPush(5);
	  d.rPush(6);
	  d.lPush(3);
	  d.lPush(2);
	  d.lPush(1);
	  while (!d.isEmpty())
      {
          System.out.print(d.lPop()+" ");
      }
  }
}

File: Stack-Queue-LinkedList/Deque/Deque.py
# complexity O(1), Amortized O(1)
class Deque:
  def __init__(self):
    self.stack1 = []
    self.stack2 = []
    self.stack3 = []

  def l_append(self, n):
    self.stack1.append(n)

  def r_append(self, n):
    self.stack2.append(n)

  def l_pop(self):
    if len(self.stack1) == 0 and len(self.stack2) != 0:
      self.__balance_left()
      
    if len(self.stack1):
      return self.stack1.pop()
    else:
      if len(self.stack2) == 1:
        return self.stack2.pop()
      else:
        return None

  def r_pop(self):
    if len(self.stack2) == 0 and len(self.stack1) != 0:
      self.__balance_right()

    if len(self.stack2):
      return self.stack2.pop()
    else:
      if len(self.stack1) == 1:
        return self.stack1.pop()
      else:
        return None

  def l_peek(self):
    if len(self.stack1) == 0:
      self.__balance_left()

    value = self.stack1.pop()
    self.stack1.append(value)
    return value

  def r_peek(self):
    if len(self.stack2) == 0:
      self.__balance_right()
    
    value = self.stack2.pop()
    self.stack2.append(value)
    return value

  def is_empty(self):
    return len(self.stack1) == 0 and len(self.stack2) == 0

  def __balance_left(self):
    n = len(self.stack2)

    while len(self.stack2):
      if len(self.stack2)  > n/2:
        self.stack3.append(self.stack2.pop())
      else:
        self.stack1.append(self.stack2.pop())

    while len(self.stack3):
      self.stack2.append(self.stack3.pop())

  def __balance_right(self):
    n = len(self.stack1)

    while len(self.stack1):
      if len(self.stack1) > n/2:
        self.stack3.append(self.stack1.pop())
      else:
        self.stack2.append(self.stack1.pop())
    while len(self.stack3):
      self.stack1.append(self.stack3.pop())


d = Deque()

d.r_append(5)
d.r_append(6)
d.r_append(7)
d.r_append(8)
d.l_append(4)
d.l_append(3)
d.l_append(2)
d.l_append(1)

while not d.is_empty():
  print('should print 1 to 8 ', str(d.l_pop()) + " ")

d.l_append(4)
d.l_append(3)
d.l_append(2)
d.l_append(1)


print('should right pop 4: ', d.r_pop())